### ISSUE TITLE
getFactoryFirxtureURL() returns null on the Factory Fixture Agent

### ISSUE THREAD

**Issue by deldrid1** [electricimp/FactoryTools#4]
I am using the FactoryTools class for developing Factory FW.  

Developing Factory FW in the Factory Environment can be a bit inconvenient for a few reasons (need to use Production logging, have to do a Promote on the Factory Model, navigate to the Production Model, do a deploy which also affects Factory Fixtures on the Production Line), so I end up having some code like what is shown below to hack around these issues and then power cycle my Device Under Test to trick it into thinking it has just received a Factory BlinkUp.

``` squirrel
g_FactoryFixtureURL <- null


if (FactoryTools.isFactoryFirmware()) {
    server.log("Running in the factory environment")
    // Running in the Factory Environment
    g_FactoryFixtureURL = FactoryTools.getFactoryFixtureURL()
} else {
    server.log("This firmware is not running in the Factory Environment");

    //Setup some default values in case we are the factory fixture, but are in a dev environment
    g_FactoryFixtureURL <- "https://agent.electricimp.com/YOUR_AGENT_ID_HERE"; 
}
```

Later on I have some code like this

``` squirrel
// A more correct set of logic is:
// if (FactoryTools.isFactoryImp()) {
//  configureFactoryImp();
//} else if (FactoryTools.isDeviceUnderTest()) {
//  configureDeviceUnderTest();
//} else {
//  server.log("Not running in Factory Environment");
//}
// but the code below is more convenient for testing.

if(g_FactoryFixtureURL == http.agenturl()){
    configureFactoryFixtureAgent()
} else {
    configureFactoryDUTAgent()
}
```

Unfortunately, getFactoryFixtureURL() currently returns null on the Factory Agent.  This was unexpected to me and I think the class should be refactored as follows:

``` squirrel
class FactoryTools {

    static version = [2,0,0];   //Unfortunately, this is a major version update since we are changing the behavior of getFactoryFixtureURL on the DUT

    /**
    * @return {bool} - 'true' or 'false'
    */
    static function isFactoryFirmware() {
        return ("factoryfirmware" in imp.configparams && imp.configparams["factoryfirmware"]);
    }

    /**
    * @return {bool}
    */
    static function isFactoryImp() {
        if ( _isAgent() ) {
            return ( isFactoryFirmware() && getFactoryFixtureURL() == http.agenturl() );
        } else {
            return ( isFactoryFirmware() && "factory_imp" in imp.configparams && imp.configparams.factory_imp == imp.getmacaddress() );
        }
    }

    /**
    * @return {bool}
    */
    static function isDeviceUnderTest() {
        return (isFactoryFirmware() && !isFactoryImp());
    }

    /**
    * Factory Device: @return {null}
    * Factory Agent: @return {null | string} - if in running on an agent the Factory Fixture agent URL is returned, otherwise null
    */
    static function getFactoryFixtureURL() {
        if ( _isAgent() ) {
            return ("factory_fixture_url" in imp.configparams) ? imp.configparams.factory_fixture_url : http.agenturl();
        }
    }

    /**
    * @return {bool}
    */
    function _isAgent() {
        return (imp.environment() == ENVIRONMENT_AGENT);
    }
}
```

Thoughts?

**Comment by betzrhodes** [comment]
It makes sense that getFactoryFixtureURL() should return the fixture's url when running on either the DUT or Factory Fixture agent.  I've done a code update to address this.  It is now available on the v2.0.0 branch, and I'll post an update when the version is available in the library system.

**Comment by betzrhodes** [comment]
v2.0.0 released.
